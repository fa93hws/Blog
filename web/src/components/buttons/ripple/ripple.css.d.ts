// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'gridBaseline': string;
  'monoWhite': string;
  'ripple': string;
  'rippleEffect': string;
  'useAnimation': string;
}
export const cssExports: CssExports;
export default cssExports;
