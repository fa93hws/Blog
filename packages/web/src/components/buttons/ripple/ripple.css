@value gridBaseline from '~components/metrics.css';
@value colorBlue from '~components/colors.css';
@value monoWhite from '~components/colors.css';

.ripple {
  position: relative;
  border-radius: 50%;
  width: calc(gridBaseline * 0.5);
  height: calc(gridBaseline * 0.5);
  opacity: 0;
  animation: rippleEffect 500ms 1;
}

.white {
  background-color: monoWhite;
}

.blue {
  background-color: colorBlue;
}

@keyframes rippleEffect {
  0% {
    transform: scale(1);
    opacity: 0.4;
  }
  100% {
    transform: scale(100);
    opacity: 0;
  }
}
