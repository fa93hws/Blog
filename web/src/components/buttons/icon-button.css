@value gridBaseline from '~components/metrics.css';

@value lengthSmall: calc(gridBaseline * 4);
@value lengthLarge: calc(gridBaseline * 8);
@value animationDuration: 800ms;

.baseButton {
  position: relative;
  overflow: hidden;
}

.iconButton {
  composes: baseButton;

  background-color: transparent;
  border: none;
  cursor: pointer;
  display: inline-flex;
  justify-content: center;
  align-items: center;
}

.iconButton.small {
  width: lengthSmall;
  height: lengthSmall;
}

.iconButton.large {
  width: lengthLarge;
  height: lengthLarge;
}

.circle {
  border-radius: 50%;
}
