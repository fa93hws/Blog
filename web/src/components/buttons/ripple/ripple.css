@value gridBaseline from '~components/metrics.css';
@value monoWhite from '~components/colors.css';

.ripple {
  position: absolute;
  background: monoWhite;
  border-radius: 50%;
  width: calc(gridBaseline * 0.5);
  height: calc(gridBaseline * 0.5);
  pointer-events: none;
}

.ripple {
  opacity: 0.4;
  transform: scale(10);
}

.ripple.useAnimation {
  opacity: 0;
  animation: rippleEffect 1500ms 1;
}

@keyframes rippleEffect {
  0% {
    transform: scale(1);
    opacity: 0.4;
  }
  100% {
    transform: scale(100);
    opacity: 0;
  }
}
