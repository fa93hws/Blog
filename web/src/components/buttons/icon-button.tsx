import * as React from 'react';
import * as classnames from 'classnames';
import { RipplableProps, withRipple } from './ripple/ripple';
import styles from './icon-button.css';

type BaseButtonProps = {
  onClick?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  onMouseDown?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  circle?: boolean;
};

type IconButtonProps = BaseButtonProps & {
  Icon: React.ComponentType;
  size: 'small' | 'large';
};

const IconButtonStateless = React.memo(
  (props: IconButtonProps & RipplableProps) => (
    <button
      onClick={props.onClick}
      onMouseDown={props.onMouseDown}
      className={classnames(styles.iconButton, styles[props.size], {
        [styles.circle]: props.circle,
      })}
    >
      <props.Icon />
      <props.RippleSlot />
    </button>
  ),
);

export const IconButton = withRipple(IconButtonStateless);
