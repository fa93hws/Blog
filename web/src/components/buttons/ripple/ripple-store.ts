import { observable, action } from 'mobx';
import { RipplePosition } from './ripple';

export class RippleStore {
  @observable.ref
  ripplePosition: RipplePosition | undefined;

  @action
  setRipple(position: RipplePosition) {
    this.ripplePosition = position;
  }
}
