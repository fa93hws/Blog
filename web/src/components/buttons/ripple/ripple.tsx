import * as React from 'react';
import { observer } from 'mobx-react';
import * as classnames from 'classnames';
import styles from './ripple.css';
import { RippleStore } from './ripple-store';

export type RipplePosition = {
  top: number;
  left: number;
};
export type RippleColor = 'white' | 'blue';

export namespace Internal {
  export const Ripple = React.memo(
    ({ position, color }: { position: RipplePosition; color: RippleColor }) => (
      <div
        key={Math.random()}
        className={classnames(styles.ripple, styles[color])}
        style={{ top: position.top, left: position.left }}
      />
    ),
  );
}

export type RipplableProps = {
  onMouseDown(e: React.MouseEvent<HTMLElement, MouseEvent>): void;
  RippleSlot: React.ComponentType<{ color: RippleColor }>;
};

const defaultStoreFactory = () => new RippleStore();
export function withRipple<T extends RipplableProps>(
  Container: React.ComponentType<T>,
  storeFactory: () => RippleStore = defaultStoreFactory,
) {
  return React.memo((originalProps: Omit<T, keyof RipplableProps>) => {
    const store = React.useMemo(storeFactory, []);
    const onMouseDown = React.useCallback(
      (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        const { left, top } = e.currentTarget.getBoundingClientRect();
        const offsetY = e.clientY - top;
        const offsetX = e.clientX - left;
        store.setRipple({ top: offsetY, left: offsetX });
      },
      [],
    );

    const RippleImpl = React.useMemo(
      () =>
        observer(({ color }: { color: RippleColor }) =>
          store.ripplePosition != null ? (
            <Internal.Ripple position={store.ripplePosition} color={color} />
          ) : null,
        ),
      [],
    );

    return (
      <Container
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        {...(originalProps as any)}
        onMouseDown={onMouseDown}
        RippleSlot={RippleImpl}
      />
    );
  });
}
