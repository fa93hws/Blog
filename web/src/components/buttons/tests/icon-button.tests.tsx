import * as React from 'react';
import { IconButton } from '../icon-button';
import { RippleStore } from '../ripple/ripple-store';

describe('IconButton', () => {
  const Icon = () => <span>icon</span>;

  it('renders a small initial IconButton', () => {
    expect(<IconButton size="small" Icon={Icon} />).toMatchRenderedSnapshot();
  });

  it('renders a large initial IconButton', () => {
    expect(<IconButton size="large" Icon={Icon} />).toMatchRenderedSnapshot();
  });

  it('renders a large circle initial IconButton', () => {
    expect(
      <IconButton size="large" circle Icon={Icon} />,
    ).toMatchRenderedSnapshot();
  });

  it('renders a large circle initial IconButton with ripple', () => {
    const store = new RippleStore();
    store.setRipple({ top: 100, left: 200 });
    expect(
      <IconButton size="large" circle Icon={Icon} store={store} />,
    ).toMatchRenderedSnapshot();
  });
});
