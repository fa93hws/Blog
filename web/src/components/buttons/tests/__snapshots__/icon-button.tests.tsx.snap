// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton renders a large circle initial IconButton 1`] = `
<button
  class="iconButton large circle"
>
  <span>
    icon
  </span>
</button>
`;

exports[`IconButton renders a large circle initial IconButton with ripple 1`] = `
<button
  class="iconButton large circle"
>
  <span>
    icon
  </span>
  <div
    class="ripple useAnimation"
    style="top:100px;left:200px"
  />
</button>
`;

exports[`IconButton renders a large initial IconButton 1`] = `
<button
  class="iconButton large"
>
  <span>
    icon
  </span>
</button>
`;

exports[`IconButton renders a small initial IconButton 1`] = `
<button
  class="iconButton small"
>
  <span>
    icon
  </span>
</button>
`;
